### ISSUE TITLE
score function estimator

### ISSUE THREAD

**Issue by yqNLP** [AlCorreia/HILAM#2]
In the paper, 
>  Practically, it allows the complete model to be trained by gradient descent to minimize a surrogate cost function defined as equation(8)

And in the code, 
`tf.reduce_mean(tf.multiply((cost-past_cost), log_probability))`
`tf.assign(past_cost, past_cost*0.9 + 0.1*tf.reduce_mean(cost))`

Can you explain the past_cost? I am new to the optimization problem with random variable and curious about it.

**Comment by AlCorreia** [comment]
That is a good question!

First of all, if you are interested in optimization of random variables, I highly recommend [this paper](https://arxiv.org/pdf/1506.05254.pdf), that discusses stochastic computation graphs as we used in our work. 

Now to answer your question, it might be helpful to think of a Reinforcement Learning framework. In that case, the feature selector is basically a **policy** that tells us which features to choose given an example from the dataset. We can learn this policy by what is called **policy gradient**, which means that instead of optimizing the classification loss (``cost`` in the code) directly, we optimize the classification loss multiplied by the log-probability of observing the chosen features, according to the current policy. That is why we talk about a "surrogate loss function" in the paper.

The intuition here is that if we make an error with a set of features that was highly unlikely according to our model, the loss should not be very high as the model already considers those features to be bad for that example, and the policy needs to be updated only slightly. Conversely, if we make an error with features that were considered very likely, then there is something wrong with the policy and we need a large gradient to update it.

The problem with policy gradient methods in general is that they have very high variance, which hampers convergence. To improve on that, we can use a baseline to reduce the variance and that is essentially what ``past_cost`` is doing there. It is an exponential moving average of past classification losses that we subtract from the current loss. For a discussion on why that works, I recommend [this other paper](https://papers.nips.cc/paper/4264-analysis-and-improvement-of-policy-gradient-estimation.pdf).

I hope that helps.

**Comment by yqNLP** [comment]
Thanks a lot. I will read through these papers. : )
